From 8877ab33b3f9cae540ae8356ded8fc9fd41f355a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Daniel=20Althviz=20Mor=C3=A9?= <d.althviz10@uniandes.edu.co>
Date: Wed, 21 Dec 2022 11:15:50 -0500
Subject: [PATCH] Workaround Qt bug on Windows with fullscreen mode in some
 screen resolutions/scaling configurations (#5401)

* Workaround Qt bug on Windows with fullscreen mode

* Improve screen detection and update comments
---
 napari/_qt/qt_main_window.py | 59 +++++++++++++++++++++++++++++++++++-
 1 file changed, 58 insertions(+), 1 deletion(-)

diff --git a/napari/_qt/qt_main_window.py b/napari/_qt/qt_main_window.py
index b720bd5c..fb1e1fcc 100644
--- a/napari/_qt/qt_main_window.py
+++ b/napari/_qt/qt_main_window.py
@@ -125,6 +125,8 @@ class _QtMainWindow(QMainWindow):
         self.setWindowTitle(self._qt_viewer.viewer.title)
 
         self._maximized_flag = False
+        self._fullscreen_flag = False
+        self._normal_geometry = QRect()
         self._window_size = None
         self._window_pos = None
         self._old_size = None
@@ -221,8 +223,63 @@ class _QtMainWindow(QMainWindow):
             with contextlib.suppress(ValueError):
                 inst = _QtMainWindow._instances
                 inst.append(inst.pop(inst.index(self)))
+
         return super().event(e)
 
+    def isFullScreen(self):
+        # Needed to prevent errors when going to fullscreen mode on Windows
+        # Use a flag attribute to determine if the window is in full screen mode
+        # See https://bugreports.qt.io/browse/QTBUG-41309
+        # Based on https://github.com/spyder-ide/spyder/pull/7720
+        return self._fullscreen_flag
+
+    def showNormal(self):
+        # Needed to prevent errors when going to fullscreen mode on Windows. Here we:
+        #   * Set fullscreen flag
+        #   * Remove `Qt.FramelessWindowHint` and `Qt.WindowStaysOnTopHint` window flags if needed
+        #   * Set geometry to previously stored normal geometry or default empty QRect
+        # Always call super `showNormal` to set Qt window state
+        # See https://bugreports.qt.io/browse/QTBUG-41309
+        # Based on https://github.com/spyder-ide/spyder/pull/7720
+        self._fullscreen_flag = False
+        if os.name == 'nt':
+            self.setWindowFlags(
+                self.windowFlags()
+                ^ (Qt.FramelessWindowHint | Qt.WindowStaysOnTopHint)
+            )
+            self.setGeometry(self._normal_geometry)
+        super().showNormal()
+
+    def showFullScreen(self):
+        # Needed to prevent errors when going to fullscreen mode on Windows. Here we:
+        #   * Set fullscreen flag
+        #   * Add `Qt.FramelessWindowHint` and `Qt.WindowStaysOnTopHint` window flags if needed
+        #   * Call super `showNormal` to update the normal screen geometry to apply it later if needed
+        #   * Save window normal geometry if needed
+        #   * Get screen geometry
+        #   * Set geometry window to use total screen geometry +1 in every direction if needed
+        # If the workaround is not needed just call super `showFullScreen`
+        # See https://bugreports.qt.io/browse/QTBUG-41309
+        # Based on https://github.com/spyder-ide/spyder/pull/7720
+        self._fullscreen_flag = True
+        if os.name == 'nt':
+            self.setWindowFlags(
+                self.windowFlags()
+                | Qt.FramelessWindowHint
+                | Qt.WindowStaysOnTopHint
+            )
+            super().showNormal()
+            self._normal_geometry = self.normalGeometry()
+            screen_rect = self.windowHandle().screen().geometry()
+            self.setGeometry(
+                screen_rect.left() - 1,
+                screen_rect.top() - 1,
+                screen_rect.width() + 2,
+                screen_rect.height() + 2,
+            )
+        else:
+            super().showFullScreen()
+
     def eventFilter(self, source, event):
         # Handle showing hidden menubar on mouse move event.
         # We do not hide menubar when a menu is being shown or
@@ -329,8 +386,8 @@ class _QtMainWindow(QMainWindow):
             self._qt_viewer.dockConsole.setVisible(False)
 
         if window_fullscreen:
-            self.setWindowState(Qt.WindowState.WindowFullScreen)
             self._maximized_flag = window_maximized
+            self.showFullScreen()
         elif window_maximized:
             self.setWindowState(Qt.WindowState.WindowMaximized)
 
-- 
2.40.0.windows.1
